plugins/omdtls/omdtls.c

```diff
@@ -0,0 +1,885 @@
+		}
+	} else {
+		dbgprintf("dtls_connect[%p]: disable certificate checking\n", pWrkrData);
+		net_ossl_set_ssl_verify_callback(pWrkrData->sslClient, SSL_VERIFY_NONE);
+	}
+ 
+	/* Create BIO from socket array! */
+	bio_client = BIO_new_dgram(pWrkrData->sockout, BIO_NOCLOSE);
+	if (!bio_client) {
+		net_ossl_lastOpenSSLErrorMsg(pData->target, 0, pWrkrData->sslClient, LOG_INFO,
+			"dtls_connect", "BIO_new_dgram");
+		ABORT_FINALIZE(RS_RET_ERR);
+	}
+	BIO_ctrl(bio_client, BIO_CTRL_DGRAM_SET_CONNECTED, 0, &pWrkrData->dtls_client_addr);
+	SSL_set_bio(pWrkrData->sslClient, bio_client, bio_client);
+
+	/* Set debug Callback for conn BIO as well! */
+	net_ossl_set_bio_callback(bio_client);
+
+	/* Blocking socket */
+//	BIO_set_nbio( bio_client, 0 );
+//	SSL_set_bio(pWrkrData->sslClient, bio_client, bio_client);
+//	SSL_set_connect_state(pWrkrData->sslClient); /*sets ssl to work in client mode.*/
+
+	dbgprintf("dtls_connect[%p]: Starting DTLS session ...\n", pWrkrData);
+	/* Perform handshake */
+	iErr = SSL_connect(pWrkrData->sslClient);
+	if (iErr <= 0) {
+		net_ossl_lastOpenSSLErrorMsg(pData->target, iErr, pWrkrData->sslClient, LOG_INFO,
+			"dtls_connect", "SSL_connect");
+		ABORT_FINALIZE(RS_RET_ERR);
+	}
+
+	// Print Cipher info
+	sslCipher = SSL_get_current_cipher(pWrkrData->sslClient);
+	dbgprintf("dtls_connect[%p]: Cipher Version: %s Name: %s\n", pWrkrData,
+		SSL_CIPHER_get_version(sslCipher), SSL_CIPHER_get_name(sslCipher));
+/*
+	// Print Peer Certificate info
+	X509 *cert = SSL_get_peer_certificate(pWrkrData->sslClient);
+	if (cert != NULL) {
+		char *line;
+		line = X509_NAME_oneline(X509_get_subject_name(cert), 0, 0);
+		dbgprintf("dtls_connect[%p]: Subject: %s\n", pWrkrData, line);
+		OPENSSL_free(line);
+	
+		line = X509_NAME_oneline(X509_get_issuer_name(cert), 0, 0);
+		dbgprintf("dtls_connect[%p]: Issuer: %s\n", pWrkrData, line);
+		OPENSSL_free(line);
+		X509_free(cert);
+	} else {
+		dbgprintf("dtls_connect[%p]: No certificates.\n", pWrkrData);
+	}
+*/
+
+	/* Set reference in SSL obj */
+	SSL_set_ex_data(pWrkrData->sslClient, 0, NULL);
+	SSL_set_ex_data(pWrkrData->sslClient, 1, NULL);
+
+	/* Set and activate timeouts */
+	struct timeval timeout;
+	timeout.tv_sec = 3;
+	timeout.tv_usec = 0;
+	BIO_ctrl(bio_client, BIO_CTRL_DGRAM_SET_RECV_TIMEOUT, 0, &timeout);
+
+
+	pWrkrData->bIsConnected = 1;
+finalize_it:
+	DBGPRINTF("dtls_connect[%p]: Exit with %d for %s:%s\n", pWrkrData, iRet, pData->target, pData->port);
+	if(iRet != RS_RET_OK) {
+		pWrkrData->bIsConnected = 0;
+	}
+	RETiRet;
+}
+
+/*
+*	Setup socket and establish virtual Connect on a SOCKET Level for DTLS 
+*/
+static rsRetVal
+dtls_init(wrkrInstanceData_t *pWrkrData)
+{
+	DEFiRet;
+	int iErr;
+	instanceData *const pData = (instanceData *const) pWrkrData->pData;
+	struct addrinfo *addrResolved = NULL;
+	struct addrinfo hints;
+	DBGPRINTF("dtls_init[%p]: setup for %s:%s\n", pWrkrData, pData->target, pData->port);
+
+	// Close Before
```

Comment:
it would be good to enhance the comment so that one knows why a close is necessary. This check in init looks a bit strange, but I guess there is a good reason for it.